File: HW/HW8/HW8.py
# Given a string s which consists of lowercase or uppercase letters, return the length of the longest palindrome that
# can be built with those letters.
#
# Letters are case-sensitive, for eresultample, "Aa" is not considered a palindrome here.
#
# Again, your problem has additional details:
#
# Constraints:
#     - 1 <= s.length <= 2000
#     - s consists of lowercase and/or uppercase English letters only.

def longestPalindrome(s: str) -> int:
    hash = {}

    counter = 0
    for i in s:
        if i in hash:
            hash.pop(i)
            counter += 1

        else:
            hash[i] = 1

    if hash:
        return counter * 2 + 1

    return counter * 2


s = "abccccdd"

longestPalindrome(s)

File: HW/HW8/HW8.java
public class HW8 {
    public static void main (String[] args) {
        String string = "abccccdd";
        longestPalindrome(string);
    }

    public static int longestPalindrome(String s) {
        int[] count = new int[128];
        for (int i = 0; i < s.length(); i++) {
            count[s.charAt(i)]++;
        }

        int result = 0;
        for (int i : count) {
            result += i - (i & 1);

            if (result % 2 == 0 && i % 2 == 1) {
                result++;
            }
        }

        System.out.println(result);
        return result;
    }
}
